reaktive/src/commonMain/kotlin/com/badoo/reaktive/observable/TakeUntilObservable.kt

```diff
@@ -0,0 +1,31 @@
+package com.badoo.reaktive.observable
+
+import com.badoo.reaktive.base.Observer
+import com.badoo.reaktive.base.subscribeSafe
+import com.badoo.reaktive.completable.CompletableCallbacks
+import com.badoo.reaktive.disposable.CompositeDisposable
+import com.badoo.reaktive.disposable.Disposable
+
+fun <T> Observable<T>.takeUntil(other: Observable<*>): Observable<T> =
+    observable {
+        val emitter = it.serialize()
+        val disposables = CompositeDisposable()
+        emitter.setDisposable(disposables)
+
+        val upstreamObserver =
+            object : ObservableObserver<T>, ObservableCallbacks<T> by emitter {
+                override fun onSubscribe(disposable: Disposable) {
+                    disposables += disposable
+                }
+            }
+
+        other.subscribeSafe(
+            object : ObservableObserver<Any?>, Observer by upstreamObserver, CompletableCallbacks by upstreamObserver {
+                override fun onNext(value: Any?) {
+                    upstreamObserver.onComplete()
```

Comment:
Disposing a disposable provided by a source on subscription is the only way to terminate the source from outside. Isn't it? Can you point me on "RxJava explicitly calls onComplete()"?

Reply:
> Can you point me on "RxJava explicitly calls onComplete()"?

Yes, here: https://github.com/ReactiveX/RxJava/blob/0c50f0ab1bcfe9bdb1c8cd13ef1dc214c405e580/src/main/java/io/reactivex/rxjava3/internal/operators/observable/ObservableTake.java#L88

Reply:
On the provided line they are calling 'downstream.onComplete', we are doing the same by calling 'upstreamObserver.onComplete' which delegates it to the downstream emitter.

Reply:
Hrm, please correct me if I'm wrong but completing upstream might have an incorrect effect when dealing with shared streams. 

Take this RxJava example:

```kotlin
val upstream = Observable.range(1, 10)
  .publish()
  .autoConnect(2)

upstream
  .take(1)
  .subscribe()

upstream
  .subscribe()
```

The second observer stays unaffected even when the first observer has completed. Is this also the expected behavior in case of Reaktive?

Reply:
In the provided example 'take' will unsubscribe from the upstream (from 'autoConnect'). But the last one will not terminate due to its nature, it will stay connected and continue working. So 'takeUntil' will produce 'onComplete' event to downstream and unsubscribe from its upstream (by disposing the disposable received in 'onSubscribe' callback). If you feel there are any bugs, please write a test case and open an issue, I will be more than happy to fix it.

Reply:
Cool cool. Just wanted to make sure I'm understanding this correctly! 👍